apps/prairielearn/src/pages/instructorAssessmentGroups/instructorAssessmentGroups.html.ts

```diff
@@ -0,0 +1,674 @@
+        <link
+          href="${nodeModulesAssetPath('tablesorter/dist/css/theme.bootstrap.min.css')}"
+          rel="stylesheet"
+        />
+        <script src="${nodeModulesAssetPath(
+            'tablesorter/dist/js/jquery.tablesorter.min.js',
+          )}"></script>
+        <script src="${nodeModulesAssetPath(
+            'tablesorter/dist/js/jquery.tablesorter.widgets.min.js',
+          )}"></script>
+      </head>
+
+      <body>
+        <script>
+          $(function () {
+            $('[data-toggle="popover"]').popover({ sanitize: false });
+
+            // Prevent the dropdown menu from closing when the popover is opened.
+            $('.js-group-action[data-toggle="popover"]').on('click', (e) => {
+              e.stopPropagation();
+            });
+
+            $('.js-group-action-dropdown').on('hide.bs.dropdown', (e) => {
+              // If the click is inside a popover, don't hide the dropdown.
+              if (e.clickEvent.target.closest('.popover')) {
+                e.preventDefault();
+                e.stopPropagation();
+                return;
+              }
+
+              // Hide all popovers when the dropdown menu is closed.
+              $('.js-group-action[data-toggle="popover"]').popover('hide');
+            });
+          });
+
+          // make the file inputs display the file name
+          $(document).on('change', '.custom-file-input', function () {
+            this.fileName = $(this).val().replace(/\\\\/g, '/').replace(/.*\\//, '');
+            $(this).parent('.custom-file').find('.custom-file-label').text(this.fileName);
+          });
+        </script>
+        ${renderEjs(__filename, "<%- include('../partials/navbar'); %>", resLocals)}
+        <main id="content" class="container-fluid">
+          ${renderEjs(
+            __filename,
+            "<%- include('../partials/assessmentSyncErrorsAndWarnings'); %>",
+            resLocals,
+          )}
+          ${typeof resLocals.errormsg !== 'undefined' && resLocals.errormsg.length > 0
+            ? html` <div class="alert alert-danger" role="alert">${resLocals.errormsg}</div> `
+            : ''}
+          ${!resLocals.isGroup
+            ? html`
+                <div class="card mb-4">
+                  <div class="card-header bg-primary text-white d-flex align-items-center">
+                    ${resLocals.assessment_set.name} ${resLocals.assessment.number}: Groups
+                  </div>
+                  <div class="card-body">
+                    This is not a group assessment. To enable this functionality, please set
+                    <code>"groupWork": true</code> in <code>infoAssessment.json</code>.
+                  </div>
+                </div>
+              `
+            : html`
+                ${resLocals.authz_data.has_course_instance_permission_edit
+                  ? html`
+                      <div class="container-fluid">
+                        <div
```

Comment:
I think moving the modals to separate functions would make this code way easier to understand.

Reply:
Moved the modals to separate functions in [39b5e80](https://github.com/PrairieLearn/PrairieLearn/pull/9550/commits/39b5e80a1ce76beee898fed6daf650f981c27704).